modules/contracts/apps/subscription/src/handlers/execute.rs

```diff
@@ -0,0 +1,351 @@
+            } else {
+                Err(SubscriptionError::NotUsingCW20Hook {})
+            }
+        }
+        SubscriptionExecuteMsg::Unsubscribe { unsubscribe_addrs } => {
+            unsubscribe(deps, env, app, unsubscribe_addrs)
+        }
+        SubscriptionExecuteMsg::ClaimEmissions { addr } => {
+            claim_subscriber_emissions(&app, &mut deps, &env, addr)
+        }
+        SubscriptionExecuteMsg::UpdateSubscriptionConfig {
+            payment_asset,
+            subscription_cost_per_week: subscription_cost,
+            subscription_per_week_emissions,
+            unsubscription_hook_addr,
+        } => update_subscription_config(
+            deps,
+            env,
+            info,
+            app,
+            payment_asset,
+            subscription_cost,
+            subscription_per_week_emissions,
+            unsubscription_hook_addr,
+        ),
+        SubscriptionExecuteMsg::RefreshTWA {} => {
+            INCOME_TWA.try_update_value(&env, deps.storage)?;
+            Ok(Response::new())
+        }
+    }
+}
+
+/// Called when either paying with a native token or through the receive_cw20 endpoint when paying
+/// with a CW20.
+pub fn try_pay(
+    app: SubscriptionApp,
+    deps: DepsMut,
+    env: Env,
+    _msg_info: MessageInfo,
+    asset: Asset,
+    subscriber_addr: Addr,
+) -> SubscriptionResult {
+    // Load all needed states
+    let config = SUBSCRIPTION_CONFIG.load(deps.storage)?;
+    let base_state = app.load_state(deps.storage)?;
+    // Construct deposit info
+    let deposit_info = config.payment_asset;
+
+    // Assert payment asset and claimed asset infos are the same
+    if deposit_info != asset.info {
+        return Err(SubscriptionError::WrongToken(deposit_info));
+    }
+    // Init vector for logging
+    let attrs = vec![
+        ("action", String::from("Deposit to subscription module")),
+        ("Received funds:", asset.to_string()),
+    ];
+
+    let maybe_subscriber = SUBSCRIBERS.may_load(deps.storage, &subscriber_addr)?;
+    // Minimum of one period worth to (re)-subscribe.
+    // prevents un- and re-subscribing all the time.
+    let required_payment = Uint128::from(DURATION_IN_WEEKS) * config.subscription_cost_per_week;
+    let paid_for_days = {
+        let paid_for_weeks = asset
+            .amount
+            .checked_div_floor(config.subscription_cost_per_week)?
+            .u128() as u64;
+        paid_for_weeks * 7
+    };
+    if let Some(mut active_sub) = maybe_subscriber {
+        // Subscriber is active, update balance
+        active_sub.expiration_timestamp = active_sub.expiration_timestamp.plus_days(paid_for_days);
+        SUBSCRIBERS.save(deps.storage, &subscriber_addr, &active_sub)?;
+    } else {
+        // Subscriber is (re)activating his subscription.
+        if asset.amount.u128() < required_payment.u128() {
+            return Err(SubscriptionError::InsufficientPayment(
+                required_payment.u128() as u64,
```

Comment:
Is this a safe cast? I guess it doesn't matter too much as it is just used in an error message but still curious.

Reply:
Yep that's unsafe